packages/components/src/Autocomplete/Autocomplete.tsx

```diff
@@ -0,0 +1,183 @@
+import { Text } from "../Text";
+import { Icon } from "../Icon";
+
+type OptionValue = string | number;
+
+export interface Option {
+  value: OptionValue;
+  label: string;
+}
+
+interface AutocompleteProps {
+  readonly initialOptions?: Option[];
+  value: Option | undefined;
+  onChange(newValue?: Option): void;
+  getOptions(newValue: string): Promise<Option[]>;
+  placeholder: string;
+}
+
+enum IndexChange {
+  Previous = -1,
+  Next = 1,
+}
+
+export function Autocomplete({
+  initialOptions = [],
+  value,
+  onChange: onSelectOption,
+  getOptions,
+  placeholder,
+}: AutocompleteProps) {
+  const [options, setOptions] = useState(initialOptions);
+  const [optionsMenuVisible, setOptionsMenuVisible] = useState(false);
+  const [text, setText] = useState((value && value.label) || "");
+  const inputRef = useRef() as MutableRefObject<InputTextRef>;
+
+  useEffect(() => {
+    if (value) {
+      setText(value.label);
+    } else {
+      setText("");
+      setOptions(initialOptions);
+    }
+  }, [value]);
+
+  return (
+    <div className={styles.autocomplete}>
+      <InputText
+        value={text}
+        onChange={textChange}
+        placeholder={placeholder}
+        onFocus={handleTextFocus}
+        onBlur={handleTextBlur}
+        ref={inputRef}
+      />
+      <Menu
+        visible={optionsMenuVisible}
+        options={options}
+        activeOption={value}
+        onOptionSelect={selectOption}
+      />
+    </div>
+  );
+
+  function selectOption(chosenOption: Option) {
+    onSelectOption(chosenOption);
+    textChange(chosenOption.label);
+    setOptionsMenuVisible(false);
+  }
+
+  async function textChange(newText: string) {
+    setOptionsMenuVisible(true);
+    setText(newText);
+    if (newText) {
+      setOptions(await getOptions(newText));
+    } else {
+      setOptions(initialOptions);
+    }
+  }
+
+  function handleTextBlur() {
+    setOptionsMenuVisible(false);
+    if (value == undefined || value.label !== text) {
+      onSelectOption(undefined);
+    }
+  }
+
+  function handleTextFocus() {
+    setOptionsMenuVisible(true);
+  }
+}
+
+interface MenuProps {
+  visible: boolean;
+  options: Option[];
+  activeOption?: Option;
+  onOptionSelect(chosenOption: Option): void;
+}
+
+function Menu({ visible, options, activeOption, onOptionSelect }: MenuProps) {
+  const [selectedIndex, setSelectedIndex] = useState(-1);
+  const optionMenuClass = classnames(styles.options, {
+    [styles.visible]: visible,
+  });
+
+  setupKeyListeners();
+
+  useEffect(() => setSelectedIndex(-1), [options]);
+
+  return (
+    <div className={optionMenuClass}>
+      {options.map((option, index) => {
+        const optionClass = classnames(styles.option, {
+          [styles.active]: index === selectedIndex,
+        });
+        return (
+          <button
+            className={optionClass}
+            key={option.value}
+            onMouseDown={handleMouseDown(option)}
```

Comment:
I'm a fan of using `bind` for currying like this.

`onOptionSelect.bind(null, option)` is equivalent to doing the thing with the extra function wrapper. The first param is the context, in this case we don't need the context.